### ISSUE TITLE
Unable to capture Unity app on Meta Quest when using page_guard

### ISSUE THREAD

**Issue by williamih** [LunarG/gfxreconstruct#1143]
**Describe the capture bug:**

When trying to capture a trace of a Vulkan-based Unity app on Meta Quest Pro with the memory tracking mode set to `page_guard`, the app gets stuck opening. After selecting the app, the OS app loading screen showing three dots is displayed, and the device stays stuck on this screen and does not progress further.

When the `gfxreconstruct` layer is not installed, or when the memory tracking mode is `assisted`, the app works fine.

Submitting this after a discussion on the Vulkan discord. I think this was your suggestion @bradgrantham-lunarg - thanks!

**Build Environment:**

Prebuilt `gfxreconstruct` binaries from https://github.com/LunarG/gfxreconstruct/releases/tag/sdk-1.3.250.0

**Capture Settings:**

"debug.gfxrecon.memory_tracking_mode" = "page_guard"
"debug.gfxrecon.page_guard_persistent_memory" = "true"
"debug.gfxrecon.page_guard_align_buffer_sizes" = "true"
"debug.gfxrecon.page_guard_unblock_sigsegv" = "true"
"debug.gfxrecon.page_guard_signal_handler_watcher" = "true"

"debug.gfxrecon.capture_file" is set to a file in "/sdcard/Download".

**To Reproduce:**

This was first observed on an internal build of an app I work on -- however, the issue happens straight after opening the app, so I expect this issue would happen with any Vulkan-based Unity app using the same driver, Unity version, etc.

Unity settings:
* Unity version 2021.2.16f1
* Oculus XR plugin version 3.0.2
* XR plugin management version 4.2.1

The layer was installed using the following method:
1. `adb push libVkLayer_gfxreconstruct.so /data/local/tmp`
2. `adb shell run-as com.company.app cp /data/local/tmp/libVkLayer_gfxreconstruct.so .`
3. `adb shell settings put global enable_gpu_debug_layers 1`
4. `adb shell settings put global gpu_debug_app com.company.app`
5. `adb shell settings put global gpu_debug_layers "VK_LAYER_LUNARG_gfxreconstruct"`
(where `com.company.app` is replaced with the APK identifier of the app)

**System environment:**

 - Meta Quest Pro (Snapdragon XR2 Plus Gen 1 platform with Qualcomm Adreno 650 GPU)
 - Vulkan version 1.1.128
- Quest system version 50000390062500330
- Firmware version 53.0.0.147.132

** Additional details:**

From the logs, it looks like SIGSEGV is being generated when internal Unity code (`vk::ImageManager::UploadThreaded()`) accesses Vulkan memory and is being treated by the OS as a crash.

[gfxreconstruct_log_except.txt](https://github.com/LunarG/gfxreconstruct/files/11658331/gfxreconstruct_log_except.txt)

Thanks!

**Comment by bradgrantham-lunarg** [comment]
Hi, @williamih !  Can you upload the logcat with the gfxrecon logs in them?  That's what I was hoping to see but there don't seem to be any in there.

**Comment by bradgrantham-lunarg** [comment]
It may be that the lack of `gfxrecon` lines in the logcat means that gfxrecon isn't being loaded.  Does a trace file get created anywhere?

**Comment by williamih** [comment]
Thanks @bradgrantham-lunarg - sorry, was just trying to save time parsing through a lot of logs. There are `gfxrecon` lines, but (at least to me) I'm not seeing anything obviously wrong.
 
[gfxreconstruct_logs_quest.txt](https://github.com/LunarG/gfxreconstruct/files/11658470/gfxreconstruct_logs_quest.txt)

**Comment by williamih** [comment]
I can also confirm that a trace file is being created, although I'd think it doesn't contain much, since the app doesn't finish starting up

**Comment by williamih** [comment]
A couple of further findings:

1. The `unassisted` memory tracking option works correctly on Quest; however, the performance impact is large and unfortunately is too slow to use the app normally.
2. I tried to work around (1) by using `debug.gfxrecon.capture_android_trigger` and only setting this to true once I've moved to the location in the app where the freeze generally occurs. However, it seems this option is not responding at runtime -- when I enable this while the app is already running, the gfxrecon trace does not capture anything or write anything to disk.
3. I also tried setting `debug.gfxrecon.capture_frames` so that the trace only starts once I'm loaded into the app, but when I set this option, nothing is traced or saved to disk.

If I use `unassisted` and don't set `debug.gfxrecon.capture_frames`, and either leave `debug.gfxrecon.capture_android_trigger` as unset or set it to true, the trace works and saves to disk, but the app performance is so slow that I can't get to the place in the app that I want to capture.

**Comment by panos-lunarg** [comment]
Hi @williamih 

I think this is clearly a case of our signal handler being replaced. The backtrace suggests that the segmentation faults takes place during a texture uploading (writing to mapped memory) and also `unassisted` method is not crashing.

The absence of the message `PageGuardManager: Signal handler has been removed. Re-installing.` from the logcat suggests that the watcher thread did get not a chance to detect the substitution before the crash happened

Nevertheless can you please try raising `debug.gfxrecon.page_guard_signal_handler_watcher_max_restores` from the default 1 to something like 2 or 3 or maybe even to something more aggressive like -1 which will effectively keep the watcher thread running? Maybe you can get the chance to get it to work

**Comment by williamih** [comment]
Hi @panos-lunarg

Thanks for looking into this - I'd missed that setting when looking at the docs originally, so it's good to know.

Oddly, when I tried to reproduce the issue today, our app has been working correctly even using `gfxrecon` with `page_guard`. The issue of getting stuck on the app opening screen isn't happening. I'm not sure what's changed here.

However, there are still some further problems.

Regarding `debug.gfxrecon.page_guard_signal_handler_watcher_max_restores`, here's what I've found:
1. Even after setting this to -1, I still don't see `PageGuardManager: Signal handler has been removed. Re-installing.` anywhere in the logs.
2. Similarly, the error messages about a crash due to SIGSEGV are still being printed even after setting that option.

Although the app is running correctly with `page_guard` enabled now, it seems the trace is not being written out correctly. There are two `.gfxr` files being generated, one of which is 0 bytes and the other is only 4 KB, even after staying in the app for several minutes. Do you know why this might be?

Thanks!

**Comment by panos-lunarg** [comment]
That's odd. Can you please attach an unabridged logcat of a "successful" run?

**Comment by williamih** [comment]
Thanks @panos-lunarg

Logs here: 
[logs-gfxrecon-infinite-restores.txt](https://github.com/LunarG/gfxreconstruct/files/11702692/logs-gfxrecon-infinite-restores.txt)

**Comment by williamih** [comment]
That xHook library looks interesting - thanks for the suggestion @panos-lunarg.

I have a couple of other thoughts/questions:

1. It would be fairly easy to modify our Vulkan code to use `vkFlushMappedMemoryRanges()` so that we could then use the `assisted` memory tracking mode. However, part of the app's rendering (mostly UI elements, etc) is handled by Unity, and we can't modify Unity's Vulkan code, so some modifications to memory mapped by Unity would be missed by `gfxreconstruct`. Could this cause corruption of the generated trace file or other errors, if the Unity-related writes were missed?

2.  As mentioned a bit earlier - I couldn't get the `debug.gfxrecon.capture_android_trigger` and `debug.gfxrecon.capture_frames` settings to work. Do you have any thoughts on how to investigate this?

**Comment by panos-lunarg** [comment]
> 1. It would be fairly easy to modify our Vulkan code to use vkFlushMappedMemoryRanges() so that we could then use the assisted memory tracking mode. However, part of the app's rendering (mostly UI elements, etc) is handled by Unity, and we can't modify Unity's Vulkan code, so some modifications to memory mapped by Unity would be missed by gfxreconstruct. Could this cause corruption of the generated trace file or other errors, if the Unity-related writes were missed?

Yes this will cause corruption. Imagine you are uploading a texture but gfxr fails to notice that, so the texture's content are not dumped into the capture file. When you replay the file, gfxr will not update the device memory so the texture will be filled with uninitialized data.

> 2. As mentioned a bit earlier - I couldn't get the debug.gfxrecon.capture_android_trigger and debug.gfxrecon.capture_frames settings to work. Do you have any thoughts on how to investigate this?

IIRC in order to use `debug.gfxrecon.capture_android_trigger` you have to set it to false when you start capturing and then set it to `true` to trigger trimming (then `false` again to stop). If it isn't set at all at the beginning then you'll be capturing as normal (non trimmed). I prepared a patch that should now print to logcat when this changes values (#1147)
It's strange that `debug.gfxrecon.capture_frames`  isn't doing anything either. Can you see anything in logcat? Gfxr should print something there when capturing trimmed even before you enter the trim range.

**Comment by williamih** [comment]
Thanks for the info - makes sense. With `debug.gfxrecon.capture_android_trigger`, I've been setting it to `false` before starting the app, then setting it to `true` once the app is running. However, nothing is captured. I can see from the logs that gfxr is detecting the `false` value when the app starts up, but it's not detecting the later modification to `true`.

I can try both options again tomorrow to double check.

**Comment by panos-lunarg** [comment]
BTW doing a trimmed capture is not going to make the crash go away. `page_guard` mechanism needs to keep track of mapped device memory while doing trimmed captures, even outside the frame-range / trigger-regions.

**Comment by panos-lunarg** [comment]
I started experimenting with integrating xHook sources in GFXR in patch #1150 
I couldn't get it to work on desktop but it did work on Android with Unreal (Fortnite).
Please give it a try and let me know it it works for you too.

**Comment by williamih** [comment]
Thanks @panos-lunarg - I'll the xHook patch out!

I was thinking of using a trimmed capture with `unassisted` memory tracking as I thought that might reduce some of the performance hit until I get to the part of the app that I want to trace. But based on your explanation of how `page_guard` works, I'd assume `unassisted` also still does all the tracking even with trimmed captures, and so there probably wouldn't be any performance benefit there?

**Comment by panos-lunarg** [comment]
Ah right. I would expect `unassisted` to be much faster outside the trim regions as it will not be dumping the mapped regions to the capture file. So it should be worth a try if for example you only want to capture a few frames from a specific scene and you don't care too much about precision.

**Comment by williamih** [comment]
Makes sense. I still can't get the `debug.gfxrecon.capture_frames` setting working unfortunately. When I set this option, nothing seems to be captured (no files written). Any thoughts on this?

**Comment by williamih** [comment]
Also - to test with the xHook patch (https://github.com/LunarG/gfxreconstruct/pull/1150), is there a way to get the Android binary from the CI build, or should I just build from source locally?

**Comment by bradgrantham-lunarg** [comment]
> Makes sense. I still can't get the `debug.gfxrecon.capture_frames` setting working unfortunately. When I set this option, nothing seems to be captured (no files written). Any thoughts on this?

Does it work if you set to "1-100"?  What are you setting it to now?

**Comment by panos-lunarg** [comment]
I was about to suggest the same as @bradgrantham-lunarg: What are you setting `debug.gfxrecon.capture_frames` to? If you request a very high frame number, depending on the application's frame rate, it can take a considerable time to reach it, and until it is reached GFXR will not create a capture file. Try something low to make sure that it works

Choosing the correct frame range can be a tough guessing game so I would suggest trying `debug.gfxrecon.capture_android_trigger` as you have a better control of when to start capturing. I just merged #1147 which should  make visible in logcat when`capture_android_trigger` changes values.

Regarding #1150, yes it would require you to build GFXR locally. I don't think the CI builds are visible externally so If you are not comfortable with building it locally I can upload it for you here.

**Comment by williamih** [comment]
Oddly, when I run `adb shell "setprop debug.gfxrecon.capture_frames '1-100'"` (following your idea @bradgrantham-lunarg) it works fine.

But when I run `adb shell "setprop debug.gfxrecon.capture_frames '2-100'"` nothing gets traced or saved to a file. Logs here for the 2-100 frame range which isn't working:

[logs-2-100.txt](https://github.com/LunarG/gfxreconstruct/files/11769090/logs-2-100.txt)

**Comment by williamih** [comment]
I'll try building myself as that's probably useful to be able to do, but can let you know if I run into any issues - thanks.

**Comment by panos-lunarg** [comment]
Interesting. Seems like we are missing the starting frame. Drafted a PR (#1153) to address it

**Comment by bradgrantham-lunarg** [comment]
> Oddly, when I run `adb shell "setprop debug.gfxrecon.capture_frames '1-100'"` (following your idea @bradgrantham-lunarg) it works fine, i.e. saves a trace file.

Your logs indicate that the "2-100" range was detected, but I don't understand how the logic in the code could miss the beginning of the range.  I'd have to suspect a strange Android behavior for strings or some such or maybe even a memory stomp.

Sorry to keep bombarding you with requests for logs.  Could you send the logs for the successful run for 1-100?

**Comment by bradgrantham-lunarg** [comment]
I think that because you're running an XR app on Quest, you're never getting a QueuePresentKHR, so we're never incrementing our frame counter.  Thus we never enter frame 2.  @davidd-lunarg , is this app a good candidate for testing out https://github.com/LunarG/gfxreconstruct/pull/1138 ?

**Comment by davidd-lunarg** [comment]
> is this app a good candidate for testing out https://github.com/LunarG/gfxreconstruct/pull/1138 ?

Yes, this may be a good app to test with #1138. If the VR app is not using vkQueuePresentKHR then GFXR will not detect frame boundaries and the entire capture will treated as frame 1 by GFXR. Note that to test #1138 the android property `debug.vr.profiler` must be set to `1` for capture.

As to why trim frames 1-100 produces a capture file but 2-100 does not, this may be due to the fact that trim ranges starting at frame 1 are treated more like [full captures](https://github.com/LunarG/gfxreconstruct/blob/8364d132a4c605bee2be66bd68a1ed66d72dd08f/framework/encode/capture_manager.cpp#L328-L338)--there is no need to track and write a trim state block. However if vkQueuePresentKHR is not being called I would guess the capture does not stop at frame 100. Instead it will capture until the app is closed.

**Comment by williamih** [comment]
Sorry for the delayed reply - not sure if/when I'll get a chance to test out those PRs as there's a few other work priorities I need to focus on right now. Thanks for the help investigating - this does seem like a really useful tool so hopefully I can come back to this sometime and get it going!

**Comment by Zydiii** [comment]
Same here [> debug.gfxrecon.capture_android_trigger](https://github.com/LunarG/gfxreconstruct/issues/1143#issuecomment-1589009448)  Could you tell me how do you make `debug.gfxrecon.capture_android_trigger` work?
